contracts/PillarStaking.sol

```diff
@@ -0,0 +1,289 @@
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IPillarStakingToken} from "./IPillarStakingToken.sol";
+
+/// @title PillarStaking
+/// @author Luke Wickens <luke@pillarproject.io>
+/// @notice staking contract for PLR tokens
+
+contract PillarStaking is ReentrancyGuard, Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 public stakingToken;
+    IERC20 public rewardToken;
+    IPillarStakingToken public stakedToken;
+    address[] public stakeholderList;
+    mapping(address => Stakeholder) public stakeholderData;
+
+    uint256 private constant BPS = 10000;
+    uint256 private constant STAKING_PERIOD = 10 days;
+    uint256 private constant STAKED_PERIOD = 52 weeks;
+    uint256 public minStake = 10000e18;
+    uint256 public maxStake = 250000e18;
+    uint256 public maxTotalStake = 7200000e18;
+    uint256 public totalStaked;
+    uint256 stakingPeriod;
+    uint256 stakedPeriod;
+    uint256 rewards;
+
+    struct Stakeholder {
+        bool staked;
+        uint256 stakedAmount;
+        uint256 rewardAmount;
+        bool claimed;
+    }
+
+    enum StakingState {
+        INITIALIZED,
+        STAKEABLE,
+        STAKED,
+        READY_FOR_UNSTAKE
+    }
+    StakingState state;
+
+    /* ========== EVENTS ========== */
+
+    event Staked(address indexed user, uint256 amount);
+    event Unstaked(address indexed user, uint256 amount);
+    event RewardsDeposited(uint256 rewards);
+    event RewardAllocated(address user, uint256 amount);
+    event RewardPaid(address indexed user, uint256 reward);
+    event ContractStateUpdated(StakingState newState);
+    event MinStakeAmountUpdated(uint256 newMinStake);
+    event MaxStakeAmountUpdated(uint256 newMaxStake);
+
+    /* ========== MODIFIERS ========== */
+
+    modifier whenInitialized() {
+        if (state != StakingState.INITIALIZED) revert OnlyWhenInitialized();
+        _;
+    }
+
+    modifier whenStakeable() {
+        if (state != StakingState.STAKEABLE) revert OnlyWhenStakeable();
+        _;
+    }
+
+    modifier whenStaked() {
+        if (state != StakingState.STAKED) revert OnlyWhenStaked();
+        _;
+    }
+
+    modifier whenReadyForUnstake() {
+        if (state != StakingState.READY_FOR_UNSTAKE)
+            revert OnlyWhenReadyForUnstake();
+        _;
+    }
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _stakingToken,
+        address _rewardToken,
+        address _stakedToken,
+        uint256 _maxTotalStake
+    ) {
+        if (_stakingToken == address(0)) revert InvalidStakingToken();
+        if (_rewardToken == address(0)) revert InvalidRewardToken();
+        if (_stakedToken == address(0)) revert InvalidStakedToken();
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        stakedToken = IPillarStakingToken(_stakedToken);
```

Comment:
Shouldn't we be instantiating a new StakingToken here rather than using an already deployed token?

Given that you are using AccessControl, if we are to deploy the contract from the PillarStaking contract then only the contract can mint/burn the tokens, no?

Reply:
No problem I can integrate that into the staking contract

Reply:
Done.